Error when an (image) upload’s mime type is unsupported

adamsilverstein opened the pull request:
<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

<!-- Insert a description of your changes here -->

Trac ticket: https://core.trac.wordpress.org/ticket/61167


<!-- insert a link to the WordPress Trac ticket here -->

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

adamsilverstein committed:
allowedMimeTypes: filter types that are not supported by the server.

adamsilverstein committed:
Cleanup

adamsilverstein committed:
Use Yoda consitionals, I must

adamsilverstein committed:
Merge branch 'trunk' into ticket/61167

adamsilverstein committed:
Merge branch 'trunk' into ticket/61167

adamsilverstein committed:
revert previous version

joedolson committed:
Customize: Accessibility: Restore heading semantics in sections.

Remove `role="presentation"` from headings in the customizer. These were needed to avoid confusing semantics when the headings also acted as buttons to control accordions, changed in [59924].

Change responsive CSS to use `screen-reader-text` styling rather than `display: none` so that mobile retains the headings hierarchy.

Props joedolson, hbhalodia, mikinc860, guillaumeturpin, rcreators, tirth03, dhrumilk . 
Fixes #62215.

git-svn-id: https://develop.svn.wordpress.org/trunk@59825 602fd350-edb4-49c9-b593-d223f7449a82

adamsilverstein committed:
Error when an (image) upload’s mime type is unsupported

adamsilverstein committed:
phpcbf

SergeyBiryukov committed:
Coding Standards: Use strict comparison in `wp_get_comment_status()`.

Follow-up to [546], [2258], [5666], [47219], [47808].

Props aristath, poena, afercia, SergeyBiryukov.
See #62279.

git-svn-id: https://develop.svn.wordpress.org/trunk@59826 602fd350-edb4-49c9-b593-d223f7449a82

SergeyBiryukov committed:
Coding Standards: Use strict comparison in `pingback()`.

Follow-up to [2983], [38852].

Props aristath, poena, afercia, SergeyBiryukov.
See #62279.

git-svn-id: https://develop.svn.wordpress.org/trunk@59827 602fd350-edb4-49c9-b593-d223f7449a82

swissspidy commented on src/wp-includes/rest-api/endpoints/class-wp-rest-attachments-controller.php:
```diff
@@ -135,6 +135,23 @@ public function create_item_permissions_check( $request ) {
+			if ( ! wp_image_editor_supports( array( 'mime_type' => $files['file']['type'] ) ) ) {
+				return new WP_Error(
+					'rest_upload_image_type_not_supported',
+					__( 'The web server cannot generate responsive image sizes for this image. Convert it to JPEG or PNG before uploading.', 'gutenberg' ),
```
Wrong text domain

adamsilverstein replied:
good catch, thanks!

adamsilverstein replied:
fixed in https://github.com/WordPress/wordpress-develop/pull/8322/commits/595a95d55f22205e92c0c5ccde24936b4740a5dc

swissspidy commented on src/wp-includes/rest-api/endpoints/class-wp-rest-attachments-controller.php:
```diff
@@ -135,6 +135,23 @@ public function create_item_permissions_check( $request ) {
+			str_starts_with( $files['file']['type'], 'image/' )
+		) {
+			// Check if the image editor supports the type.
+			if ( ! wp_image_editor_supports( array( 'mime_type' => $files['file']['type'] ) ) ) {
```
Could a test be added for this?

adamsilverstein replied:
Yes, probably. I will give it a try.

adamsilverstein replied:
added test in https://github.com/WordPress/wordpress-develop/pull/8322/commits/eb41d06d00f62def6520d5a12b8b1aef10cdca25 - I am currently unable to test this locally as my travel setup supports AVIF, hopefully tests work in CI.

swissspidy commented on src/wp-includes/rest-api/endpoints/class-wp-rest-attachments-controller.php:
```diff
@@ -135,6 +135,23 @@ public function create_item_permissions_check( $request ) {
 			);
 		}
 
+		// If the upload is an image, check if the server can handle the mime type.
```
Could we add a way to disable this behavior, e.g. with a filter?

With client-side image processing I need to be able to upload AVIF images and the like without any additional server-side processing.

adamsilverstein replied:
That is a good suggestion, I was wondering about the same thing. If we do add a filter, I feel like it should impact both editor and media uploads, what do you think?

swissspidy replied:
That makes sense to me, yes.

johnbillion committed:
Security: Switch to using bcrypt for hashing user passwords and BLAKE2b for hashing application passwords and security keys.

Passwords and security keys that were saved in prior versions of WordPress will continue to work. Each user's password will be opportunistically rehashed and resaved when they next subsequently log in using a valid password.

The following new functions have been introduced:

* `wp_password_needs_rehash()`
* `wp_fast_hash()`
* `wp_verify_fast_hash()`

The following new filters have been introduced:

* `password_needs_rehash`
* `wp_hash_password_algorithm`
* `wp_hash_password_options`

Props ayeshrajans, bgermann, dd32, deadduck169, desrosj, haozi, harrym, iandunn, jammycakes, joehoyle, johnbillion, mbijon, mojorob, mslavco, my1xt, nacin, otto42, paragoninitiativeenterprises, paulkevan, rmccue, ryanhellyer, scribu, swalkinshaw, synchro, th23, timothyblynjacobs, tomdxw, westi, xknown.

Additional thanks go to the Roots team, Soatok, Calvin Alkan, and Raphael Ahrens.

Fixes #21022, #44628

git-svn-id: https://develop.svn.wordpress.org/trunk@59828 602fd350-edb4-49c9-b593-d223f7449a82

SergeyBiryukov committed:
Docs: Correct formatting for `get_the_date` and `get_the_time` DocBlocks.

Follow-up to [53000].

See #62281.

git-svn-id: https://develop.svn.wordpress.org/trunk@59829 602fd350-edb4-49c9-b593-d223f7449a82

swissspidy committed:
General: Fix `force_ssl_admin()` to always return bool.

Props pbearne, costdev, autotutorial, debarghyabanerjee, swissspidy.
Fixes #60023.

git-svn-id: https://develop.svn.wordpress.org/trunk@59830 602fd350-edb4-49c9-b593-d223f7449a82

johnbillion committed:
Security: Remove use of `innerHTML` in the `the_block_template_skip_link()` function.

There is no need to support HTML in this string and switching to `innerText` helps facilitate a more restrictive Content Security Policy.

Props micromadness, sabernhardt

Fixes #58765

git-svn-id: https://develop.svn.wordpress.org/trunk@59831 602fd350-edb4-49c9-b593-d223f7449a82

adamsilverstein committed:
Test uploading unsupported image type

adamsilverstein committed:
remove namespace copied from gutenberg PR

adamsilverstein committed:
Merge branch 'trunk' into ticket/61167

adamsilverstein committed:
phpcbf

adamsilverstein committed:
ensure file type set before using

adamsilverstein committed:
Add and test new wp_prevent_unsupported_image_uploads filter

adamsilverstein committed:
Add wp_prevent_unsupported_image_uploads filter to media.php

adamsilverstein committed:
Cleanup

adamsilverstein commented:
I added a filter to disable the behavior both in the REST API and in media.php where we set up the upload limitation for the media library. Along the way I noticed we aren't handling support for HEIC images in the core check which might be worth adding in a future ticket.

@swissspidy - appreciate any feedback here on the filter naming and the approach.

adamsilverstein committed:
phpcbf

adamsilverstein committed:
Merge branch 'trunk' into ticket/61167

adamsilverstein committed:
Merge branch 'ticket/61167' of github.com:adamsilverstein/wordpress-develop into ticket/61167

adamsilverstein committed:
phpcbf

swissspidy commented on src/wp-includes/rest-api/endpoints/class-wp-rest-attachments-controller.php:
```diff
@@ -135,6 +135,32 @@ public function create_item_permissions_check( $request ) {
+		/**
+		 * Filter whether the server should prevent uploads for image types it doesn't support. Default true.
+		 *
+		 * Developers can use this filter to enable uploads of certain image types.
```
Missing `@since` tag

adamsilverstein replied:
added in cdb9d8b9f7ee7a53a5f690d85cb8661a8dd9f52a

swissspidy commented on src/wp-includes/rest-api/endpoints/class-wp-rest-attachments-controller.php:
```diff
@@ -135,6 +135,32 @@ public function create_item_permissions_check( $request ) {
+		 *
+		 * @param bool $check_mime Whether to prevent uploads of unsupported image types.
+		 */
+		$prevent_unsupported_uploads = apply_filters( 'wp_prevent_unsupported_image_uploads', true );
```
I think we should pass the current mime type as a second argument. That gives developers more flexibility.

swissspidy replied:
Another thought: if in the long term we want to do support this for other mime types too, we can ditch the `_image_` from the filter name or use `_mime_type_`, e.g. `wp_prevent_unsupported_mime_type`. WDYT?

adamsilverstein replied:
👍🏼 `wp_prevent_unsupported_mime_type` -  I like it, makes sense to pass the mime type so developers can act on that, and the name should reflect that.

adamsilverstein replied:
Adding this here, but it will be null in `media_upload_form` because we don't know the mime type yet when the filter is run.

I wound up using `wp_prevent_unsupported_mime_type_uploads` as the name since this specifically applies to preventing uploads and this helps clarify that.

adamsilverstein committed:
Merge branch 'trunk' into ticket/61167

adamsilverstein committed:
Add missing `@since` tag

adamsilverstein committed:
Add mime type as a second parameter and rename filter

adamsilverstein committed:
update doc block

adamsilverstein commented:
@swissspidy I have addressed your feedback and this is ready for another review.

swissspidy reviewed (approved):
Looks good to me now 👍

adamsilverstein commented:
Closing as fixed in https://core.trac.wordpress.org/changeset/60084
